examples/source-to-image/assemble

```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+
+set -x
+
+set -eo pipefail
+
+# Remove any 'environment.yml' or 'requirements.txt' files which may
+# have been carried over from the base image so we don't reinstall
+# packages which have already been installed. This could occur where
+# an S2I build was used to create a new base image with pre-installed
+# Python packages, with the new image then subsequently being used as a
+# S2I builder base image.
+
+rm -f /home/$NB_USER/work/environment.yml
+rm -f /home/$NB_USER/work/requirements.txt
+
+# Copy injected files to correct place in 'work' directory.
+
+cp -Rf /tmp/src/. /home/$NB_USER/work
```

Comment:
Okay, I thought it must have been some existing convention that the work directory existed and that people were expected to always do stuff in it for some reason. I could see that as reasonable as mounting onto ``$HOME`` actually causes problems where as part of the image build things were created/cached in ``$HOME`` that would be required later. Mounting on ``$HOME`` would hide any such cached files and potentially break things or cause things to be regenerated. As an example, mounting on ``$HOME`` would hide ``~/.cache/matplotlib`` generated during the build of ``scipy-notebook``, meaning the font cache index (???) has to be regenerated first time ``matplotlib`` is imported in a notebook.

Having only ``$HOME`` to work with is a bit problematic in more general case of S2I because S2I builders may want to generate special configuration and startup scripts. You cannot put these in the directory which is likely going to be what people would mount a volume on as then it all stops working when you deploy the image. Because S2I doesn't run as ``root``, you don't really have any ability to create directories anywhere else really you can put stuff. The more typical scenario in S2I builders is that you have ``/opt/app-root`` under which the S2I builder can put stuff, not a user home directory, and just ``/opt/app-root/src`` is ``$HOME`` for the user account, and is what is used by the application. Thus S2I builders can happily write extra stuff under ``/opt/app-root`` and know that things will not be broken if single subdirectory of ``/opt/app-root/src`` underneath is used as mount point for volume.

If am to ignore the ``work`` directory, and possibly even start removing it since no need for it these examples, then means I have to generate some configuration files elsewhere, perhaps under ``/etc/jupyter`` since that is writable to ``users`` group.

Reply:
We used to put required configuration, kernelspecs, etc. in $HOME and that caused us to call 'work' the place we wouldn't touch where users could put their files and could be safely mounted as a volume.  We've since fixed that, and by design no files in $HOME may be required for any of these images to work. In JupyterHub, the current convention is to mount $HOME as a volume, because then there's no confusion about users losing data because they forgot to put it in their 'persistent' directory.

The only data in $HOME in any of these images is the matplotlib font cache, which is a small performance optimization for the first import of matplotlib. If $HOME is mounted as a volume, the cache will be created as normal for any user and re-used thereafter. So it saves an every-time cost for the no-persistence case and persistent storage saves the same cost when there is persistent storage, after the first launch for that user.

The exception is Binder, where we put the repo data is in $HOME specifically because it's an ephemeral container, and we don't need to persist user data. The same makes sense if s2i is being used for ephemeral containers, but doesn't if the users should be saving their work in $HOME. If s2i should be writing persistent data for the home-as-volume case, *nothing* should be written to $HOME. Writing to /opt or /etc makes sense. Whatever you like.